### PR TITLE
feat: add non-disruptive esxi cert replacement

### LINKED ISSUE #72: Add support to include the ESXi host management IP in the CSR for `SubjectAltName`
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [x] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [x] Do not leave "+1" or other comments that do not add relevant information or questions.
- [x] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

Currently, when generating Certificate Signing Requests (CSRs) for ESXi hosts using the `Request-EsxiCsr` function, there is no support for including the management IP Address of the ESXi host in the Subject Alternative Name (`SubjectAltName`) field of the CSR. 

This request proposes the enhancement of the `Request-EsxiCsr` function to include this, as an optional parameter that can be used. 

### Use Case(s)

- To align the process to be more in line with what is present in SDDC Manager today, which gives hints for Subject Alternative Name creation as part of the certificate generation workflow. 

### Potential Configuration

- `New-VIMachineCertificateSigningRequest` already includes support for SubjectAltName.
- `Get-VMhost` can be used to obtain the IP address of the management VMkernel adapter, in a similar way the value of `$esxiHosts` is retrieved.
- Could include `-IncludeSubjectAltNameIp` parameter in `Request-EsxiCsr` function to make it optional.

### References

[`New-VIMachineCertificateSigningRequest`](https://developer.vmware.com/docs/powercli/latest/vmware.powercli.vcenter/commands/new-vimachinecertificatesigningrequest/#VCenter)

### LINKED ISSUE #120: Add support for non-disruptive ESXi certificate replacement in vcf-5.2
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

Enhanced ESXi certificate replacement for VCF 5.2, to take advantage of the enhancement (ESXi Non-disruptive certificate replacement) in vSphere 8.0.3.

### Use Case(s)

To take advantage of the enhancement (ESXi Non-disruptive certificate replacement) in vSphere 8.0.3, to shorten the time and complexity involved when performing ESXi certificate replacement.

### Potential Configuration

None, will add version check and split procedure calls off the existing commands for current cmdlet.

### References

### PR DESCRIPTION
**Pull request by garlicNova**
### Summary

Adds support to `Install-VcfCertificate` for non-disruptive ESXi certificate replacement.  

### Type

- [ ] Bugfix
- [x] Enhancement or Feature
- [ ] Code Style or Formatting
- [ ] Documentation
- [ ] Refactoring
- [ ] Chore
- [ ] Other
        Please describe:

### Breaking Changes?

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### Test and Documentation

- [x] Tests have been completed.
- [ ] Documentation has been added or updated.

### Issue References

Closes #120 
Closes #72

### CODE REVIEW

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Instead of a v2, can you update `Install-EsxiCertificate` to check the version and if >= 5.2.0 use this path, but if < 5.2.0 use the prior so that we don't add confusion with a v2.

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 1]
```diff
@@ -1401,9 +1418,42 @@ Function Install-EsxiCertificate {
...
+                    Write-Error "Private key not found at $keyPath. Skipping certificate replacement for ESXi host $esxiFqdn."
+                    $skippedHosts.Add($esxiFqdn)
+                    continue
+                }
+            }
+
             if (Confirm-EsxiCertificateInstalled -server $server -user $user -pass $pass -esxiFqdn $esxiFqdn -signedCertificate $crtPath) {
                 $skippedHosts.Add($esxiFqdn)
                 continue
+            } elseif ($vcfVersion -eq "5.2") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.CertificateManagement.psm1:1433]: -ge ?

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 2]
```diff
@@ -1401,9 +1418,42 @@ Function Install-EsxiCertificate {
                 continue
             }
 
+            if ($vcfVersion -eq "5.2") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.CertificateManagement.psm1:1421]: -ge ?

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 3]
```diff
@@ -1378,13 +1381,27 @@ Function Install-EsxiCertificate {
             if (!$esxiHosts) { Write-Error "No ESXi host $esxiFqdn found in workload domain $domain." -ErrorAction Stop }
         }
 
-        # Perform ESXi host vSAN data migration pre-check.
-        if ($PsBoundParameters.ContainsKey("cluster")) {
-            Write-Output "Performing Data Migration Pre-check on the cluster $cluster"
-            Get-EsxiHostVsanMaintenanceModePrecheck -server $server -user $user -pass $pass -domain $domain -cluster $cluster -vsanDataMigrationMode $vsanDataMigrationMode
+        $version = Get-VCFManager -version
+        $vcfVersion = $version.Split('.')[0]+"."+$version.Split('.')[1]
+
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.CertificateManagement.psm1:1386]: -ge ?

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 4]
```diff
@@ -2207,13 +2257,28 @@ Function Install-VCFCertificate {
 
         .EXAMPLE
         Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -esxiFqdn sfo01-m01-esx01.sfo.rainpole.io -migratePowerOffVMs -vsanDataMigrationMode EnsureAccessibility -certificateDirectory F:\certificates -certificateFileExt ".cer"
-        This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path.  The ESXi host
-        will enter maintenance mode with Migrate Power off VMs option enabled and vSAN data migration Mode set to EnsureAccessibility.
+        This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path.  When VMware Cloud Foundation 
+        version is earlier than 5.2, the ESXi host will enter maintenance mode with Migrate Power off VMs option enabled and vSAN data migration Mode set to EnsureAccessibility.
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.CertificateManagement.psm1:2261]:
```suggestion
        version is earlier than 5.2, the ESXi host will enter maintenance mode with migrate powered off VMs option enabled and vSAN data migration mode set to EnsureAccessibility.
```

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 5]
```diff
@@ -2278,44 +2347,109 @@ Function Install-VCFCertificate {
...
-        } elseif ($PSBoundParameters.ContainsKey("cluster") -and $PSBoundParameters.ContainsKey("esxiFqdn")) {
-            Write-Error "Only one of -esxiFqdn or -cluster parameter can be provided at a time."
-        } elseif ($PSBoundParameters.ContainsKey("cluster")) {
-            if ($vsanDataMigrationMode.IsPresent) {
-                if ($migratePowerOffVMs.IsPresent) {
-                    Install-EsxiCertificate -server $server -user $user -pass $pass -domain $domain -cluster $cluster -certificateDirectory $certificateDirectory -certificateFileExt $certificateFileExt -vsanDataMigrationMode $vsanDataMigrationMode -migratePowerOffVMs
+        $version = Get-VCFManager -version
+        $vcfVersion = $version.Split('.')[0]+"."+$version.Split('.')[1]
+
+        if ($vcfVersion -eq "5.2") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.CertificateManagement.psm1:2363]: -ge ?

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 6]
```diff
@@ -52,15 +52,36 @@ This example will connect to SDDC Manager to install the signed certificates for
 Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -esxiFqdn sfo01-m01-esx01.sfo.rainpole.io -migratePowerOffVMs -vsanDataMigrationMode EnsureAccessibility -certificateDirectory F:\certificates -certificateFileExt ".cer"
 ```
 
-This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path. The ESXi host will enter maintenance mode with Migrate Power off VMs option enabled and vSAN data migration Mode set to `EnsureAccessibility`.
+This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path. When VMware Cloud Foundation 
+version is earlier than 5.2, the ESXi host will enter maintenance mode with Migrate Power off VMs option enabled and vSAN data migration Mode set to `EnsureAccessibility`.
 
-### Example 3
+### EXAMPLE 3
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:58]:
```suggestion
### Example 3
```

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 7]
```diff
@@ -52,15 +52,36 @@ This example will connect to SDDC Manager to install the signed certificates for
...
 ```powershell
-Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user <administrator@vsphere.local> -pass VMw@re1! -domain sfo-m01 -cluster sfo-m01-cl01 -vsanDataMigrationMode EnsureAccessibility -certificateDirectory F:\certificates -certificateFileExt ".cer"
+Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -cluster sfo-m01-cl01 -certificateDirectory F:\certificates -certificateFileExt ".cer"
 ```
 
-This example will install certificates for each ESXi host in cluster sfo-m01-cl01 in workload domain sfo-m01 from the provided path. The ESXi host will enter maintenance mode with Migrate Power off VMs option disabled and vSAN data migration Mode set to `EnsureAccessibility`.
+This example will install certificates for each ESXi host in cluster sfo-m01-cl01 in workload domain sfo-m01 from the provided path.  When VMware Cloud Foundation 
+version is 5.2 or later, the vsanDataMigrationMode option no longer applied.  
+
+### EXAMPLE 4
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:67]:
```suggestion
### Example 4
```

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 8]
```diff
@@ -52,15 +52,36 @@ This example will connect to SDDC Manager to install the signed certificates for
...
+
+```powershell
+Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -cluster sfo-m01-cl01 -certificateDirectory F:\certificates -certificateFileExt ".cer" -uploadPrivateKey
+```
+
+This example will install private keys and certificates for each ESXi host in cluster sfo-m01-cl01 in workload domain sfo-m01 from the provided path.  The uploadprivatekey 
+parameter is only validate for VMware Cloud Foundation version is 5.2 or later.
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:74]:
```suggestion
parameter is only available for VMware Cloud Foundation version is 5.2 or later.
```

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 9]
```diff
@@ -52,15 +52,36 @@ This example will connect to SDDC Manager to install the signed certificates for
...
+
+### EXAMPLE 5
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:76]:
```suggestion
### Example 5
```

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 10]
```diff
@@ -52,15 +52,36 @@ This example will connect to SDDC Manager to install the signed certificates for
...
+
+```powershell
+Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -esxiFqdn sfo01-m01-esx01.sfo.rainpole.io -migratePowerOffVMs -vsanDataMigrationMode EnsureAccessibility -certificateDirectory F:\certificates -certificateFileExt ".cer"
+```
+
+This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path.  When VMware Cloud Foundation
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:82]:
```suggestion
This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path. When VMware Cloud Foundation 
```

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 11]
```diff
@@ -52,15 +52,36 @@ This example will connect to SDDC Manager to install the signed certificates for
...
+parameter is only validate for VMware Cloud Foundation version is 5.2 or later.
+
+### EXAMPLE 5
+
+```powershell
+Install-VCFCertificate -esxi -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -esxiFqdn sfo01-m01-esx01.sfo.rainpole.io -migratePowerOffVMs -vsanDataMigrationMode EnsureAccessibility -certificateDirectory F:\certificates -certificateFileExt ".cer"
+```
+
+This example will install the certificate to the ESXi host sfo01-m01-esx01.sfo.rainpole.io in domain sfo-m01 from the provided path.  When VMware Cloud Foundation 
+version is earlier than 5.2, the ESXi host will enter maintenance mode with Migrate Power off VMs option enabled and vSAN data migration Mode set to EnsureAccessibility.
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:83]: This reads a bit odd.

File: docs/documentation/functions/Install-VCFCertificate.md [hunk 12]
```diff
@@ -289,6 +310,22 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -uploadPrivateKey
+
+Option to upload a custom private key when performing the ESXi host certificate replacement (only valid for VCF 5.2 and later).
```
Comment (tenthirtyam) [inline docs/documentation/functions/Install-VCFCertificate.md:315]:
```suggestion
Option to upload a custom private key when performing the ESXi host certificate replacement. Supported on VMware Cloud Foundation 5.2 or later.
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 1 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 2 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 3 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1:1383]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 4 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 5 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 6 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md:58]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 7 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md:67]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 8 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md:74]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 9 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md:76]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 10 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md:100]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

**Review by tenthirtyam — APPROVED** [review]
LGTM!

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 12 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md]: addressed in commit e6a7844965249faebca841b4da521bef3b7b40ba

File: docs/documentation/functions/Install-VCFCertificate.md (hunk 11 above)
Comment (garlicNova) [reply docs/documentation/functions/Install-VCFCertificate.md:101]: pending will update this in a later.

**Pull request merged**